### PR TITLE
Reset cross-db flag for each statement

### PR DESCRIPTION
**Pull request by HarshLunagariya**
### Description

Earlier, while parsing the batch, We were marking the all the subsequent statements of the batch as cross-database statements once we encounter any cross-database statement in batch which was incorrect.

This commit fixes above behaviour by resetting 'is_cross_db' maintained at mutator level once we mark some statement as cross_db in the parse tree.

### Issues Resolved

BABEL-5447

### Test Scenarios Covered ###
* **Use case based -**

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by shalinilohia50 — APPROVED** [review]
LGTM!

File: test/JDBC/expected/BABEL-CROSS-DB-vu-verify.out [hunk 1]
```diff
@@ -793,6 +793,372 @@ int
 ~~END~~
 
 
+-- tsql
+-- Cross-db testcases for statements in batch/block
+USE master;
+GO
+
+-- Batches having system objects
+BEGIN; SELECT DISTINCT 1 FROM my_babel_cross_db_vu_prepare_db1.dbo.sysdatabases; BEGIN; SELECT DISTINCT 1 FROM babel_cross_db_vu_prepare_master_t1; END; END;
```
Comment (rishabhtanwar29) [inline test/JDBC/expected/BABEL-CROSS-DB-vu-verify.out:802]: 1. Let's also add test where we call `sys` objects as well like db1.sys.databases etc.
2. Create temp tables after executing cross-db stmt as DDL permissions work differently.

Comment (anju15bharti) [reply test/JDBC/expected/BABEL-CROSS-DB-vu-verify.out:802]: Also let's include below cases:
1. tests for statements in batch inside a transaction.
2. some cases where login/user has some default db and schemas with it and it's executing one cross-db query with other queries in a batch.

**Review by anju15bharti — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp [hunk 2]
```diff
@@ -1917,7 +1917,10 @@ class tsqlBuilder : public tsqlCommonMutator
 
 		// record whether stmt is cross-db
 		if (is_cross_db)
+		{
 			stmt->is_cross_db = true;
+			is_cross_db = false;
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp:1923]: We are also setting this flag to true in `makeExecuteProcedure` but not resetting it back. What would happen if we exec a batch like
```
EXEC db.dbo.p
SELECT * FROM t
```

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp (hunk 2 above)
Comment (tanscorpio7) [reply contrib/babelfishpg_tsql/src/tsqlIface.cpp:1923]: can we reset is_cross_db in enterDMLStatement ?

Comment (tanscorpio7) [reply contrib/babelfishpg_tsql/src/tsqlIface.cpp:1923]: Also seems like we can just remove use of this flag from `makeExecuteProcedure`

Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/tsqlIface.cpp:1923]: > Also seems like we can just remove use of this flag from makeExecuteProcedure

Yes, I noticed it too. On top of that we can cleanup some logic from pl_exec as well. We should track it separately because I assume there we will need more test coverage, can't just rely on existing tests to conclude that given logic is redundant.

> can we reset is_cross_db in enterDMLStatement ?

Done.

File: test/JDBC/expected/BABEL-CROSS-DB-vu-verify.out (hunk 1 above)
Comment (HarshLunagariya) [reply test/JDBC/expected/BABEL-CROSS-DB-vu-verify.out:802]: > * Let's also add test where we call `sys` objects as well like db1.sys.databases etc.
> * Create temp tables after executing cross-db stmt as DDL permissions work differently.

Done.

> Also let's include below cases:
> 
> 1. tests for statements in batch inside a transaction.
> 2. some cases where login/user has some default db and schemas with it and it's executing one cross-db query with other queries in a batch.

Regarding, 2nd point - not adding anything for default db setting since it doesn't affect anything while query execution. It is only utilised while setting up a new db session.

Comment (HarshLunagariya) [reply test/JDBC/expected/BABEL-CROSS-DB-vu-verify.out:802]: regarding default_schema, It is being tracked by BABEL-5050, needs separate fix.

**Review by rishabhtanwar29 — DISMISSED** [review]

**Review by tanscorpio7 — APPROVED** [review]

**Review by KushaalShroff — DISMISSED** [review]

File: test/JDBC/expected/BABEL-CROSS-DB-vu-verify.out [hunk 3]
```diff
@@ -794,5 +794,1130 @@ int
...
+GO
+
+INSERT INTO my_babel_cross_db_vu_prepare_db1.dbo.babel_cross_db_vu_prepare_db1_t1 (a)
+SELECT a 
+FROM master.dbo.babel_cross_db_vu_prepare_master_t1;
+GO
+~~ROW COUNT: 4~~
+
+
+SELECT * FROM my_babel_cross_db_vu_prepare_db1.dbo.babel_cross_db_vu_prepare_db1_t1;
```
Comment (rishabhtanwar29) [inline test/JDBC/expected/BABEL-CROSS-DB-vu-verify.out]: This one needs an ORDER BY clause or else it will be flaky.

File: test/JDBC/expected/BABEL-CROSS-DB-vu-verify.out [hunk 4]
```diff
@@ -794,5 +794,1130 @@ int
...
+
+UPDATE my_babel_cross_db_vu_prepare_db1.dbo.babel_cross_db_vu_prepare_db1_t1
+SET a = m.a * 2
+FROM master.dbo.babel_cross_db_vu_prepare_master_t1 m
+WHERE my_babel_cross_db_vu_prepare_db1.dbo.babel_cross_db_vu_prepare_db1_t1.a = m.a;
+GO
+~~ROW COUNT: 1~~
+
+
+SELECT * FROM my_babel_cross_db_vu_prepare_db1.dbo.babel_cross_db_vu_prepare_db1_t1;
```
Comment (rishabhtanwar29) [inline test/JDBC/expected/BABEL-CROSS-DB-vu-verify.out]: This one also needs an ORDER BY clause.

File: test/JDBC/expected/BABEL-CROSS-DB-vu-verify.out [hunk 5]
```diff
@@ -794,5 +794,1130 @@ int
...
+
+INSERT INTO dbo.my_babel_cross_db_vu_prepare_db1_t2 (a)
+SELECT t1.a
+FROM master.dbo.babel_cross_db_vu_prepare_master_t1 t1
+JOIN dbo.babel_cross_db_vu_prepare_db1_t1 t2 ON t1.a = t2.a;
+GO
+~~ROW COUNT: 1~~
+
+
+SELECT * FROM dbo.my_babel_cross_db_vu_prepare_db1_t2;
```
Comment (rishabhtanwar29) [inline test/JDBC/expected/BABEL-CROSS-DB-vu-verify.out]: Same here, needs ORDER BY.

**Review by rishabhtanwar29 — APPROVED** [review]
